src/samplers/support/hmc_core.jl

```diff
@@ -75,18 +108,52 @@ end
 
 # Compute Hamiltonian
 find_H(p::Vector, model::Function, vi::VarInfo, spl::Sampler) = begin
-  # NOTE: getlogp(vi) = 0 means the current vals[end] hasn't been used at all.
-  #       This can be a result of link/invlink (where expand! is used)
-  if getlogp(vi) == 0 vi = runmodel(model, vi, spl) end
 
-  p_orig = p .* spl.info[:wum][:stds]
+  # Old code
+  # # NOTE: getlogp(vi) = 0 means the current vals[end] hasn't been used at all.
+  # #       This can be a result of link/invlink (where expand! is used)
+  # if getlogp(vi) == 0 vi = runmodel(model, vi, spl) end
+  #
+  # p_orig = p .* spl.info[:wum][:stds]
+  #
+  # H = dot(p_orig, p_orig) / 2 + realpart(-getlogp(vi))
+  # if isnan(H) H = Inf else H end
+  #
+  # H
+
+  logpdf_func_float(theta) = begin
 
-  H = dot(p_orig, p_orig) / 2 + realpart(-getlogp(vi))
-  if isnan(H) H = Inf else H end
+      vi[spl][:] = theta[:]
+      realpart(runmodel(model, vi, spl).logp)
+
+  end
+
+  return _find_H(vi[spl], p, logpdf_func_float, spl.info[:wum][:stds])
+
+end
+
+function _find_H(theta::T, p::Vector, logpdf_func_float::Function, stds::Vector) where {T<:Union{Vector,SubArray}}
+
+  lp = logpdf_func_float(theta)
+
+  return _find_H(theta, p, lp, stds)
 
-  H
 end
 
+function _find_H(theta::T, p::Vector, lp::Real, stds::Vector) where {T<:Union{Vector,SubArray}}
+
+  p_orig = p .* stds
+
+  H = 0.5 * dot(p_orig, p_orig) + (-lp)
+
+  H = isnan(H) ? Inf : H
```

Comment:
Why is this necessary? Looks like this treats the symptoms of an underlying problem, rather than addressing them directly.

Reply:
The program may give a `NaN` if any constraint applied in the variables is violated or any function inside has some invalid computation, at some point `theta` in the sampling space. To correctly sample in this situation we can simply do rejection sampling, i.e. rejecting any `theta` which gives `NaN`. Plugging rejection sampling in HMC still gives a valid algorithm. One way to do this rejection is simply setting `H` to `Inf`, in which case the MH criteria will reject the sample with a probability of 1.

This is also what Stan's implementation on such rejection: https://github.com/stan-dev/stan/blob/a79d22c3bc9dbfe002ea7174a0f9f8d690e33580/src/stan/mcmc/hmc/static/base_static_hmc.hpp#L61

Reply:
That makes sense. Thanks for the clarification, @xukai92!